src/main/java/com/mongodb/jdbc/MongoSQLResultSetMetaData.java

```diff
@@ -2,57 +2,144 @@
+        String datasource;
+        int index;
+
+        DatasourceAndIndex(String datasource, int index) {
+            this.datasource = datasource;
+            this.index = index;
+        }
+    }
+
+    // A mapping from columnLabel name to datasource name and index.
+    private Map<String, DatasourceAndIndex> columnLabels;
+    // A mapping from index position to NameSpace (datasource, columnLabel).
+    private List<NameSpace> columnIndices;
+    // A mapping from index position to ColumnTypeInfo.
+    private List<MongoSQLColumnTypeInfo> columnTypeInfo;
+
+    // This gets the datasource for a given columnLabel, and is used
+    // in MongoSQLResultSet to retrieve data by label.
+    String getDatasource(String columnLabel) {
+        return columnLabels.get(columnLabel).datasource;
+    }
+
+    private static void assertObjectSchema(MongoJsonSchema schema) throws SQLException {
+        if (schema.bsonType == null
+                || !schema.bsonType.equals("object")
+                || schema.properties == null) {
+            throw new SQLException("ResultSetMetaData json schema must be object with properties");
+        }
+    }
+
+    private void processDataSource(MongoJsonSchema schema, String datasource) throws SQLException {
+        MongoJsonSchema datasourceSchema = schema.properties.get(datasource);
+        assertObjectSchema(datasourceSchema);
+
+        Object[] columns = datasourceSchema.properties.keySet().toArray();
+        Arrays.sort(columns);
+
+        for (Object column : columns) {
+            String columnAsStr = (String) column;
+            MongoJsonSchema columnSchema = datasourceSchema.properties.get(columnAsStr);
+            columnIndices.add(new NameSpace(datasource, columnAsStr));
+            int subNullability =
+                    (datasourceSchema.required == null
+                                    || !datasourceSchema.required.contains(columnAsStr))
+                            ? ResultSetMetaData.columnNullable
+                            : ResultSetMetaData.columnNoNulls;
+            columnTypeInfo.add(new MongoSQLColumnTypeInfo(columnSchema, subNullability));
+            if (!columnLabels.containsKey(columnAsStr)) {
+                columnLabels.put(
+                        columnAsStr, new DatasourceAndIndex(datasource, columnIndices.size() - 1));
+            }
+        }
+    };
+
+    public MongoSQLResultSetMetaData(MongoJsonSchema schema) throws SQLException {
+        assertObjectSchema(schema);
+
+        columnLabels = new HashMap<String, DatasourceAndIndex>();
+        columnIndices = new ArrayList<NameSpace>();
+        columnTypeInfo = new ArrayList<MongoSQLColumnTypeInfo>();
+
+        Object[] datasources = schema.properties.keySet().toArray();
+        Arrays.sort(datasources);
+
+        for (Object datasource : datasources) {
+            String datasourceAsString = (String) datasource;
+            processDataSource(schema, datasourceAsString);
```

Comment:
```suggestion
        String[] datasources = schema.properties.keySet().toArray(new String[0]);
        Arrays.sort(datasources);

        for (string datasource : datasources) {
            processDataSource(schema, datasource);
```
same can be done here